### DISCUSSION TITLE
SymForce Internship Program!

### DISCUSSION THREAD

**Discussion by hayk-skydio** [symforce-org/symforce#242] [Announcements]
The SymForce team is excited to announce internship opportunities to students, researchers, and engineers interested in the space. Given the broad applications of symbolic computation and code generation, we are open to a diverse set of backgrounds and experience levels.

If you are interested, reach out with your CV and interest area at `hayk at skydio dot com`.

🟢 Internships are funded by Skydio
🟢 Contributions are open source
🟢 Remote work is encouraged
🟢 Timing and project scope is flexible
🟢 Interns are paired with an experienced mentor

Below is a curated set of impactful projects. Please feel free to suggest your own project ideas!

## CUDA Language Backend

Add a backend to generate CUDA code from symbolic expressions, such as the inner loop of image or volumetric processing kernels, or to create custom layers for machine learning frameworks like PyTorch or TensorFlow. This capability is widely useful and opens up SymForce to be used in more domains. This project includes creating a compelling example of this capability.

## PyTorch Integration

PyTorch dominates the ML research community, and there is increasing use of complex differentiable operations within deep network architectures. This project enables generating PyTorch functional graphs from SymForce symbolic expressions and their derivatives. This would provide a powerful combination for modern machine learning. This project includes creating a compelling example of this integration.

## SymPy Mechanics Example

The [SymPy Mechanics](https://docs.sympy.org/latest/modules/physics/mechanics/index.html) module and [PyDy](https://www.pydy.org/) provide a powerful set of tools for robotics and physics, such as the ability to symbolically constructing rigid bodies, the interactions between them, and compute their equations of motion using Lagrange’s or Kane’s method. These approaches are readily integrated with SymForce to generate fast, ready-to-use, production functions for simulating or controlling these dynamics. If you are a student, researcher, or engineer interested in SymForce for robot arm manipulation, humanoid control, or similar tasks, please reach out.

## Optimize Bundle Adjustment Performance

SymForce provides an [example](https://github.com/symforce-org/symforce/tree/main/symforce/examples/bundle_adjustment_in_the_large) to load and solve [Bundle-Adjustment-in-the-Large](https://grail.cs.washington.edu/projects/bal/) (BAL) datasets.  However, this use case is not particularly optimized for performance and there are many promising enhancements, which should generalize across use cases. Additionally, a head-to-head benchmark is needed to compare performance on this problem to Ceres and GTSAM.

## Constrained Optimization

SymForce does not currently support hard constraints in its tangent-space Levenberg-Marquardt implementation. Adding a constrained optimization solver, for example Augmented Lagrangians (although we’re open to other ideas!), would expand the capabilities of the solver. This project will involve implementing the optimizer in C++, integrating with the existing optimization system, and providing an example that demonstrates its usefulness. A related opportunity is integration with the [ProxQP](https://github.com/Simple-Robotics/proxsuite) library.

## Optimize Conditional Evaluation

This project aims to improve the ability to evaluate subsets of generated functions, depending on the outputs desired by the caller. Examples of features include:

- Not evaluating temporaries which are only necessary for some outputs, when those outputs are not requested: https://github.com/symforce-org/symforce/issues/78
- Enabling hard branches in generated code in general: https://github.com/symforce-org/symforce/issues/73

## Windows Support

SymForce is not currently supported or tested on Windows platforms, but there are no known blockers. This project aims to expand the build system to support building from source on Windows and generating Windows wheels in SymForce releases. This will likely involve modifications to our CMake build, additions to our CI to test Windows builds, and testing.

## Coverage Reporting

Add test coverage reports for our Python and C++ code and automatically generate coverage reports in CI. Previous work has been done to track Python code using [coverage.py](https://coverage.readthedocs.io/en/6.5.0/), but is not currently running. Another potential component of this project would be running the tests under various tools like ASan, UBSan, and valgrind in CI.

**Comment by simutisernestas** [comment]
Are opportunities open for people outside of US ?

**Reply by hayk-skydio** [reply]
Yes! This program is remote friendly and is meant to be a lightweight internship process focused on open-source contributions. Feel free to contact us.

**Comment by ShivamPR21** [comment]
Hi! Are these positions still taking candidates?

**Reply by aaron-skydio** [reply]
We are!  You can send your CV and interest area to me (`aaron at skydio dot com`) and Hayk
